Moves group properties into a metadata property (#89)
design-tokens/community-group: technical-reports/format/groups.md

```json
{
  "token-uno": {
    "value": "token value 1"
  },
  "token-group": {
    "token-dos": {
      "value": "token value 2"
    },

>     "nested-token-group": {

Reviewer: Nice!

  The format editors acknowledge existing best-practices for token naming, but place no direct constraints on naming via the specification.
</p>

## Additional group metadata properties

> Groups may include additional metadata properties. To minimize naming clashes with items in the group, all metadata properties are nested inside an object whose key is `metadata`. `metadata` is therefore a reserved word within groups meaning that tokens or nested groups cannot be called `metadata`.

Reviewer 2: ```suggestion
Groups MAY include additional metadata properties. To minimize naming collisions with items in the group, all metadata properties are nested inside an object whose key is `metadata`. `metadata` is therefore a reserved word within groups, meaning that tokens and nested groups MUST NOT be called `metadata`.
```

```json
{
  "brand": {
    "metadata": {
      "description": "Design tokens from our brand guidelines"

>     },

Reviewer 2: The idea behind having `description` at the same level as `value` was to make it easily accessible, and ideally nudge authors toward documenting tokens.

Nesting the `description` field deeper _could_ have a negative impact on this.

Curious if that's something that was brought up when deciding about this change?

Author: I don't recall that specific point being brought up, so thanks for raising it.

Bear in mind that this change only affects _groups_, not tokens. In tokens, `value`, `description` and other properties all remain top-level. That's because, as per the spec, no other properties are allowed within a token.

On groups, properties are used as the names for items within that group, so having "special" properties like `description` mixed in amongst them could be problematic if we ever added new group properties in future spec versions as there'd be the risk of naming collisions with item names people might be using. Moving all group properties under `metadata` avoids this.

As far as nudging folks to add descriptions to groups goes, not having `description` as a top level property on groups is of course less convenient to type. However, I doubt it's that much of a chore that it would put people of describing their group at all. With a bit of luck, IDEs may one day offer auto-completion (a la Intellisense) for editing DTCG files which would further aleviate the issue.

Arguably, having group `description` and other properties all collated into an object of their own may actually aid readability as they might otherwise be randomly mixed in amongst the group's item names.

I'm therefore still in favour of making this change. What do others think?

Reviewer 2: Thank you for this great context, this makes more sense now 👍🏻 

And my apologies, I hadn't realized this was only impacting the schema of groups: my assumption was that if group descriptions were under `metadata.description`, for consistency they'd also have been nested in that location for tokens themselves.

This then raises the point: could this nesting inconsistency make the format harder to learn?

To illustrate my question, here's an explanation:

```json5
{
  amazingGroup: {
    // group description (nested under metadata object)
    metadata: { description: "my group" },
    // token description (direct child of the token)
    fantasticToken: { value: 15, description: "my token" },

    // expectation possibly set by group.metadata.description
    fantasticToken: {
      value: 15,
      metadata: { description: "my token" },
    }
  }
}
```

Reviewer 3: I like the idea of having consistent formatting whenever possible. And looking at this example now I can see the very subtle difference. Would having `metadata` on the group and the item create the same issue as having `description` at the top love?

Reviewer 4: From a schema design perspective, it’s good to think in terms of what will break if a reserved word is added/removed/renamed. Namespacing (`metadata.*`) is a great way to mitigate against this.

Tokens are “closed,” meaning they only accept few reserved words (`type`, `value`, `description`). The potential for name changes won’t break much, as a user probably shouldn’t have any arbitrary keys on tokens. So in this scenario, `metadata` would just add length to everything when typing it out.

But groups are “open” in that a user may use any key they like. So any time a reserved word is added/removed/renamed, it may break all users’ schemas. Namespacing all reserved words into `metadata` (as the PR/Issue discuss) solves this.

All that to say it might _seem_ inconsistent on the surface to have `metadata` on one and not the other, but IMHO isn’t inconsistent in practice because tokens and groups behave differently and have different constraints, and it’s a reflection of that.

Author: Thanks @drwpow , I couldn't have said it better myself :-)

Author: For what it's worth, there was an earlier suggestion in #61 by @ivnmaksimovic to do the inverse: All group's top-level properties are "closed" - just like they are for tokens - but there is a special `tokens` property, which then contains just the child items. For example:

```json
{
  "group-1": {
    "description": "Description for group-1",
    "tokens": {
      "token-1": { "value": "..." },
      "token-2": { "value": "..." },
      "nested-group": {
        "tokens": {
          "deep-token-1": { "value": "..." }
        }
      }
    }
  }
}
```

I suppose you could argue it's more consistent from a stylistic point of view. But the trade-off is that it would require _every_ group to always contain a `tokens` property (regardless of whether or not is also uses any other special group properties like `description` or `type`). It also makes your JSON files nest more deeply which could be a bit inconvenient when hand-editing files.

There's pros and cons either way, but my personal preference is still to go ahead with `metadata` for groups only. I'm keen to hear the other format editor's views though.

Author: I've added [a comment](https://github.com/design-tokens/community-group/issues/61#issuecomment-1011544318) in the "reserved names" issue (#61) mentioning that we decided at the last format editors' meeting to debate this further before deciding how to proceed.

As I've mentioned there, I'm now wondering if we should reconsider prefixing all "special" properites in our spec (both for groups and for tokens). That also avoids naming clashes but keeps the syntax for groups and tokens similar.

```json
{
  "brand": {
    "metadata": {

>       "type": "color"

Reviewer 4: I like this. This allows someone to both set a default type for this group, as well as also create a `type` subgroup, which wouldn’t be out-of-the-question for typography tokens.

Author: Good point! i hadn't actually considered that before.

In effect, this means that only `value` and `metadata` need to be reserved words. The reason is that `value` is the only required prop in a token and therefore the way in which parsers can differentiate between a group and a token.

I.e. if an object has a property called  `value`, it must be a token. If not, ituat be a group. Consequently, groups cannot contain items called `value`.

Reviewer 5: Yeah, name clashing crossed my mind when I read this spec today, I like this change as well!

Reviewer 4: To expand on this: I’m currently working on a [parser for design tokens](https://github.com/drwpow/cobalt-ui) and the `type` key is a really important part of the schema, both for parsers as well as type systems like TypeScript to infer which is which. Having both a group and token possess a top-level `type: "color"` key would make everything exponentially harder/slower to reason about. So for that reason, I do like it being `metadata` on the group.

If group did have `type`, it’d make things easiest for parsers to be `type: "group"`. But that would probably cause confusion with the `metadata.type` proposal here, and overall that wouldn’t add _that_ much value for parsers as it can be inferred.
